_posts/2019-12-10-bert-model.md

```diff
@@ -0,0 +1,410 @@
+the model. It was a tough job, and while reading various blogs, papers, articles
+it was hard to process all that information. There are I think a lot of factors
+which I think only experienced ML people would get at first instance. I had to
+Google almost every new term I had discovered while reading. So, I spent time
+reading and understanding and after I had a basic idea, I was ready to look at
+the example code.
+
+The Transformers.jl repository already has few examples on BERT. Here we are
+working on Sentiment Analysis and hence we need to construct a classifier. We
+will now see the code and the output for the model I constructed. Its a bit
+long, and I won't be able to explain each and every section of the code. But,
+I'll cover up the important ones.
+
+### Step 1: Prepare everything
+
+First thing I believe is to prepare the data. You should download all the
+data files and keep them inside a folder. You'd need some more files but, that
+would be a part of the code itself. Also, I'm using IJulia to code everything
+out, and I would highly recommend it since it can save you a lot of time,
+especially if you are new to Julia.
+
+Also, we need to have all the dependencies installed before we move towards
+coding.
+
+>Note: You'll need to have Julia 1.1 installed to manage the dependencies.
+This might change later on, but for now Transformers and its dependencies
+might not work on other versions.
+
+```
+julia> ]
+
+(v1.1) pkg> add Transformers
+```
+
+Specific version of Flux is required here.
+```
+(v1.1) pkg> add Flux@0.9.0
+```
+
+We need CuArrays at version 1.3.0.
+```
+(v1.1) pkg> add CuArrays@1.3.0
+```
+
+Currently, TextAnalysis works with Flux 0.9.
+```
+(v1.1) pkg> add TextAnalysis
+```
+
+Mange the versions as you'd want them. These versions *can* change in the future
+and you'll need to find one combination that works with all. Once, you've done
+that we can move towards coding part.
+
+You can find entire notebook here. (No link, will publish it after GCI).
+
+First, lets get all of the packages imported
+
+```jl
+using CuArrays
+
+using Transformers
+using Transformers.Basic
+using Transformers.Pretrain
+using Transformers.Datasets
+using Transformers.BidirectionalEncoder
+
+using Flux
+using Flux: onehotbatch, gradient
+import Flux.Optimise: update!
+using WordTokenizers
+```
+
+Also, some constants we'll be using:
+
+```jl
+const labels = ("0", "1")
+const opt = ADAM(1e-6)
+const Batch = 2
+const Epoch = 1
+```
+Of course, you can change these as per your requirement. I've set Batch to 2
+to save some memory and Epoch to 1 to save some time. Epoch being one just
+would mean that there will be a single iteration of weights update.
+
+Now, lets get the pretrained model. Here the downloading extracting the model
+will be taken care of by Transformers. All you need to do is following.
+
+```jl
+const _bert_model, wordpiece, tokenizer = pretrain"Bert-uncased_L-12_H-768_A-12"
```

Comment:
explicitly mention here that loading the pretrained model gives you three things -- the model weights, the vocabulary, and the tokenizer. It's easy for the reader to miss this and be confused later